### PR TITLE
CDAP-15672. HTTP Realtime Source

### PR DESCRIPTION
**Pull request by aonischuk**
Implement Realtime source. Which supports:
- checkpointing. To continue from the last processed page
- uses InputDStream for checkpointing
- returns RDD from inputDStream by parts
- supports comparing and reloading the last page records if the page changes (hash checks). Is controlled by "Track last page changes" config.

For more details: https://wiki.cask.co/display/CE/HTTP+Realtime+Source

### CODE REVIEW

File: src/main/java/io/cdap/plugin/http/source/common/pagination/IncrementAnIndexPaginationIterator.java [hunk 1]
```diff
@@ -61,4 +65,28 @@ protected String getNextPageUrl(HttpResponse response, BasePage page) {
...
+
+  private Pattern getUrlPattern() {
+    String[] urlParts = config.getUrl().split(PAGINATION_INDEX_PLACEHOLDER_REGEX);
+
+    String regexpString = Pattern.quote(urlParts[0]) + "([\\d]+)";
+    if (urlParts.length > 1) {
+      regexpString += Pattern.quote(urlParts[1]);
+    }
+
+    return Pattern.compile(regexpString);
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/http/source/common/pagination/IncrementAnIndexPaginationIterator.java]: this operation is not cheap. Can it be compiled once?

File: src/main/java/io/cdap/plugin/http/source/streaming/HttpStreamingSourceConfig.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    "maximum is not enforced.")
+  @Macro
+  protected Long maxPagesPerFetch;
+
+  protected HttpStreamingSourceConfig(String referenceName) {
+    super(referenceName);
+  }
+
+  public Boolean shouldTrackLastPageChanges() {
+    return Boolean.parseBoolean(trackLastPageChanges);
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/http/source/streaming/HttpStreamingSourceConfig.java]: Configs do not support Boolean type?

**Review by mlozbin-cybervisiontech — CHANGES_REQUESTED** [review]

File: src/main/java/io/cdap/plugin/http/source/common/pagination/IncrementAnIndexPaginationIterator.java (hunk 1 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/IncrementAnIndexPaginationIterator.java]: done

File: src/main/java/io/cdap/plugin/http/source/streaming/HttpStreamingSourceConfig.java (hunk 2 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/streaming/HttpStreamingSourceConfig.java]: done

**Review by mlozbin-cybervisiontech — APPROVED** [review]

**Pull request merged**